onadata/apps/api/viewsets/data_viewset.py

```diff
@@ -647,8 +647,11 @@ def _get_data(self, query, fields, sort, start, limit, is_public_request):
             if should_paginate:
                 self.paginator.page_size = retrieval_threshold
 
-        if not isinstance(self.object_list, types.GeneratorType) and \
+        if isinstance(self.object_list, types.GeneratorType) and \
                 should_paginate:
+            # Unpack generator object to list
+            self.object_list = list(self.object_list)
+        else:
```

Comment:
Seems like the pagination headers are skipped if the object_list is a generator object?

Reply:
Yes.
I assume, from the previously changes I observed a scenerio whereby data returned in JSON format was to be streamed, with no pagination.
I left the else condition to continue to handle such kinds of requests, i.e that the data returned is a generator objects and we do not want to paginate the returned response data.
A good example can be found on [this](https://github.com/onaio/onadata/blob/e2ca6f770466c42773a87031a3bcf1b6966b98e4/onadata/apps/api/tests/viewsets/test_data_viewset.py#L755) test, we first return a generator object, and we dont want to paginate it, but use the limit flag instead.

Reply:
IMO we should always paginate requests if requested no matter if we're streaming the data... It's still going to be a pain if a user requests a lot of records while streaming data is enabled

Reply:
We also have a few tests that use pagination and stream data https://github.com/onaio/onadata/blob/e2ca6f770466c42773a87031a3bcf1b6966b98e4/onadata/apps/api/tests/viewsets/test_data_viewset.py#L653-L654